ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/OngoingAuthorizationObserver.scala

```diff
@@ -29,12 +63,85 @@ private[auth] final class OngoingAuthorizationObserver[A](
   override def setMessageCompression(b: Boolean): Unit = observer.setMessageCompression(b)
 
   override def onNext(v: A): Unit =
-    authorized(claims) match {
+    authorize match {
+      case _ if aborted.get() => ()
       case Right(_) => observer.onNext(v)
-      case Left(authorizationError) => observer.onError(throwOnFailure(authorizationError))
+      case Left(authorizationError: AuthorizationError) =>
+        val e: StatusRuntimeException =
+          errorFactories.permissionDenied(authorizationError.reason)(errorLogger)
+        observer.onError(e)
     }
 
   override def onError(throwable: Throwable): Unit = observer.onError(throwable)
 
   override def onCompleted(): Unit = observer.onCompleted()
+
+  private def authorize: Either[AuthorizationError, Unit] = {
+    val now = nowF()
+    for {
+      _ <- originalClaims.notExpired(now)
+      _ = validateClaimsResolvedFromUser(now)
+    } yield {
+      ()
+    }
+  }
+
+  /** Aborts the stream by throwing an exception
+    * if any change in compared the original user claims has been detected
+    * or if user claims change check times out.
+    */
```

Comment:
The code in here works with the limitation that a user rights check is only initiatiated on seeing a new stream element, which means that quiet streams won't be terminated in a timely fashion on user right changes or expiry.

If it's not too much effort, then I'd prefer to not have this problem and terminate in a timely fashion. This will require running the user rights check independently of the stream in a periodic fashion. That would also allow to adapt the tests to not require sending an extra element.

@nmarton-da : what is the recommended way to do a check periodically in Scala? If found this https://stackoverflow.com/a/39938355, but I'm sure we already have place where we do that.

Reply:
My initial question is which and/or how many threads do want to spare with these checks.

If one new thread per all streams then `java.util.Timer` looks like a good fit.
If we predict needing more than one dedicated thread then should it be a dedicated thread pool or one that we already use?

Reply:
Will probably start by looking at scheduling with Akka

Reply:
> If it's not too much effort, then I'd prefer to not have this problem and terminate in a timely fashion.
Ok